R/ebirdhotspotlist.R

```diff
@@ -27,56 +27,84 @@
-#' @examples \dontrun{
+#' @importFrom dplyr select
+#' @examples
+#' \dontrun{
 #' ebirdhotspotlist("CA-NS-HL")
 #' ebirdhotspotlist("VA")
 #' ebirdhotspotlist(lat = 30, lng = -90, dist = 10)
 #' library(httr)
 #' ebirdhotspotlist("CA-NS-HL", config = verbose())
 #' }
-#' @author Sebastian Pardo \email{sebpardo@@gmail.com}
+#' @author Sebastian Pardo \email{sebpardo@@gmail.com},
+#'    David Bradnum \email{dbradnum@@gmail.com}
 #' @references \url{http://ebird.org/}
 
-ebirdhotspotlist <-  function(regionCode = NULL, lat = NULL, lng = NULL, dist = NULL, back = NULL,
-                      sleep = 0, key = NULL, ...) {
-  
+ebirdhotspotlist <- function(regionCode = NULL, lat = NULL, lng = NULL, dist = NULL, back = NULL,
+                             sleep = 0, key = NULL, ...) {
   if (length(regionCode) > 1) {
     stop("More than one location specified")
   }
-  
-  if (is.null(regionCode) & any(is.null(lat), is.null(lng))) stop("Both region and lat/lng include NULL values")
-  
+
   if (!is.null(back)) {
     if (back > 30) {
       back <- 30
       warning("'Back' supplied was >30 days, using 30 days")
     }
     back <- round(back)
   }
-  
-  if (is.null(regionCode) & all(!is.null(lat), !is.null(lng))) {
-    message("Coordinates provided instead of region code, locating hostpots using lat/lng")
-    lat <- round(lat, 2)
-    lng <- round(lng, 2)
-    
+
+  if (is.null(regionCode)) {
+    message("No region code provided, locating hotspots using lat/lng")
+
+    geoloc <- c(lat, lng)
+
+    if (length(geoloc) == 1) {
+      stop("Incomplete lat/lng coordinates provided; please provide both values.")
+    }
+
+    if (is.null(geoloc)) geoloc <- getlatlng()
+
+    if (abs(geoloc[1]) > 90) {
+      stop("Please provide a latitude between -90 and 90 degrees.")
+    }
+
+    if (abs(geoloc[2]) > 180) {
+      stop("Please provide a longitude between -180 and 180 degrees.")
+    }
+
+    lat <- round(geoloc[1], 2)
+    lng <- round(geoloc[2], 2)
+
     if (!is.null(dist)) {
       if (dist > 500) {
         dist <- 500
         warning("Distance supplied was >500km, using 500km")
       }
       dist <- round(dist)
     }
-    
-    url <- paste0(ebase(),"ref/hotspot/geo?lat=", lat, "&lng=", lng) 
+
+    url <- paste0(ebase(), "ref/hotspot/geo?lat=", lat, "&lng=", lng)
     args <- ebird_compact(list(fmt = "json", back = back, dist = dist))
   } else {
+    if (nchar(as.character(regionCode)) == 0) {
+      stop("regionCode must not be empty")
+    }
+
     invisible(ebirdregioninfo(regionCode, key = key))
     url <- paste0(ebase(),"ref/hotspot/", regionCode)
     args <- ebird_compact(list(fmt = "json", back = back))
   }
-  
+
   Sys.sleep(sleep)
-  ebird_GET(url, args = args, key = key, ...)
+  res <- ebird_GET(url, args = args, key = key, ...)
+
+  # remove locID column which is currently returned in duplicate by the API
+  colToRemove <- "locID"
+  if (colToRemove %in% names(res)) {
+    res <- select(res, -colToRemove)
+  }
+
+  res
```

Comment:
It is usually preferable to avoid importing more functions from other packages if you can do the same thing simply in base R. I would replace lines 104 to 109 with:
```suggestion
res[ , !names(res) == "locID"]
```

This is also robust if the API removes the "locID" field in the future.

Reply:
OK, that makes sense - have amended. I'm so used to writing tidyverse-style code now that I often gloss-over base R too quickly!